transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -162,10 +163,19 @@ public void run() {
 
     private void invokeChannelRegistered() {
         if (invokeHandler()) {
-            try {
-                ((ChannelInboundHandler) handler()).channelRegistered(this);
-            } catch (Throwable t) {
-                invokeExceptionCaught(t);
+            final ChannelHandler handler = handler();
+            if (handler instanceof ChannelDuplexHandler) {
+                try {
+                    ((ChannelDuplexHandler) handler).channelRegistered(this);
+                } catch (Throwable t) {
+                    invokeExceptionCaught(t);
+                }
+            } else {
+                try {
+                    ((ChannelInboundHandler) handler).channelRegistered(this);
+                } catch (Throwable t) {
+                    invokeExceptionCaught(t);
+                }
             }
```

Comment:
```
            final ChannelHandler handler = handler();
            if (handler instanceof ChannelDuplexHandler) {
                try {
                    ((ChannelDuplexHandler) handler).channelRegistered(this);
                } catch (Throwable t) {
                    invokeExceptionCaught(t);
                }
            } else {
                try {
                    ((ChannelInboundHandler) handler).channelRegistered(this);
                } catch (Throwable t) {
                    invokeExceptionCaught(t);
                }
            }
```

What if you do:

```
        final ChannelHandler handler = handler();
        try {
            if (handler instanceof ChannelDuplexHandler) {
                ((ChannelDuplexHandler) handler).channelRegistered(this);
            } else {
                ((ChannelInboundHandler) handler).channelRegistered(this);
            }
        } catch (Throwable t) {
            invokeExceptionCaught(t);
        }
```

Less bytecode usually is better.

Reply:
Good point, privately @chrisvest already pointed the same :)
I have prepared the pr in hurry to let folks to try it if they have a perf rig.

Reply:
how it looks now @normanmaurer ?